pkg/kube/servicewatcher.go

```diff
@@ -0,0 +1,149 @@
+	informerFactory := informers.NewSharedInformerFactory(client, 1*time.Hour)
+	serviceInformer := informerFactory.Core().V1().Services()
+	sharedInformer := serviceInformer.Informer()
+	sharedInformer.AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc: func(obj interface{}) {
+			handleUpdate(nil, obj, eventCh)
+		},
+		DeleteFunc: func(obj interface{}) {
+			handleUpdate(obj, nil, eventCh)
+		},
+		UpdateFunc: func(oldObj, newObj interface{}) {
+			handleUpdate(oldObj, newObj, eventCh)
+		},
+	})
+	err := sharedInformer.SetWatchErrorHandler(func(r *cache.Reflector, err error) {
+		log.Debugw("kubernetes: error watching", log.Fields{
+			"error": err,
+		})
+		switch {
+		case apierrors.IsResourceExpired(err):
+			// resource expired; the informer will adapt, ignore.
+		case apierrors.IsGone(err):
+			// resource is gone; the informer will adapt, ignore.
+		case apierrors.IsServiceUnavailable(err):
+			// service unavailable; it should come back later.
+		case errors.Is(err, io.EOF):
+			// watch closed normally; ignore.
+		case errors.Is(err, io.ErrUnexpectedEOF):
+			// connection interrupted; informer will retry, ignore.
+		case isTimeout(err):
+			// connection is a time out of some sort, this is fine
+		case errors.Is(err, unix.ECONNREFUSED):
+			// connection refused; the server is down.
+			// Note that "failed to list" errors need k8s.io/client-go 0.25.0
+			errorCh <- err
+		default:
+			var statusError *apierrors.StatusError
+			if errors.As(err, &statusError) {
+				log.Debugw("kubernetes: got status error", log.Fields{
+					"status": statusError.Status(),
+					"debug":  fmt.Sprintf(statusError.DebugError()),
+				})
+			}
+			log.Errorw("kubernetes: unexpected error watching", log.Fields{
+				"error": err,
+			})
+		}
+	})
+	if err != nil {
+		return nil, nil, fmt.Errorf("error watching services: %w", err)
+	}
+	informerFactory.WaitForCacheSync(ctx.Done())
+	informerFactory.Start(ctx.Done())
+	services, err := serviceInformer.Lister().List(labels.Everything())
+	if err != nil {
+		return nil, nil, fmt.Errorf("error listing services: %w", err)
+	}
+	// List the initial set of services asynchronously, so that we don't have to
+	// worry about the channel blocking.
+	go func() {
+		for _, svc := range services {
+			handleUpdate(nil, svc, eventCh)
+		}
+	}()
+	return eventCh, errorCh, nil
+}
+
+// handleUpdate examines the old and new services, calculating the difference
+// and emitting events to the given channel.
+func handleUpdate(oldObj, newObj interface{}, eventCh chan<- event) {
+	deleted := make(map[int32]struct{})
+	added := make(map[int32]struct{})
+	oldSvc, _ := oldObj.(*corev1.Service)
+	newSvc, _ := newObj.(*corev1.Service)
+	namespace := "<unknown>"
+	name := "<unknown>"
+
+	if oldSvc != nil {
+		namespace = oldSvc.Namespace
+		name = oldSvc.Name
+		if oldSvc.Spec.Type == corev1.ServiceTypeNodePort {
+			for _, port := range oldSvc.Spec.Ports {
+				deleted[port.NodePort] = struct{}{}
+			}
+		}
+	}
```

Comment:
That would be difficult, since we don't want to assign to `namespace`/`name` when it's `nil` (e.g. when adding/removing, instead of updating).  That means we can basically reuse ~4 lines, which doesn't seem worth it.

Reply:
I did reduce repetition around the code to emit events, though.